File: python/space-age/space_age.py
from enum import Enum
from functools import partial

BASE = 31557600 # base earth orbital period

class Planet(Enum):
  EARTH   = BASE
  MERCURY = BASE * 0.2408467
  VENUS   = BASE * 0.61519726
  MARS    = BASE * 1.8808158
  JUPITER = BASE * 11.862615
  SATURN  = BASE * 29.447498
  URANUS  = BASE * 84.016846
  NEPTUNE = BASE * 164.79132

  def __init__(self, orbital_period):
    self.orbital_period = orbital_period

  def space_age(self, seconds):
    return round(seconds / self.orbital_period, 2)

class SpaceAge(object):
  def __init__(self, seconds):
    self.seconds = seconds

    for planet in Planet:
      method_name = "on_{}".format(planet.name.lower())
      setattr(self, method_name, partial(self.on, planet))

  def on(self, planet):
    return planet.space_age(self.seconds)

File: go/space-age/space_age.go
package space

import "fmt"

const secondsInYear = 31557600

type Planet string

// Given an age in seconds,
// calculate how old someone is on specified planet
func Age(seconds float64, planet Planet) float64 {
	return earthAge(seconds) / orbitalPeriod(planet)
}

// converts seconds to earth years
func earthAge(seconds float64) float64 {
	return seconds / secondsInYear
}

// determines the orbital period for a planet
func orbitalPeriod(planet Planet) float64 {
	switch planet {
	case "Earth":
		return 1
	case "Mercury":
		return 0.2408467
	case "Venus":
		return 0.61519726
	case "Mars":
		return 1.8808158
	case "Jupiter":
		return 11.862615
	case "Saturn":
		return 29.447498
	case "Uranus":
		return 84.016846
	case "Neptune":
		return 164.79132
	default:
		panic(fmt.Sprintf("Unknown planet %s", planet))
	}
}
